src/internal/collection/README.md

```diff
@@ -0,0 +1,103 @@
+ * `TTL` - get the time-to-live of a given item in the collection
+ * `ListRev` - stream the entire collection alongside the etcd revision of each item
+ * `WatchByIndex` - watch the entire collection based on an index (unused)
+
+### ReadWrite
+
+A `ReadWriteCollection` is constructed with a transaction (`STM` for etcd, `sqlx.Tx` for postgres) which is used to isolate any operations until the transaction is finished.  The transaction itself may reattempt its callback multiple times in case the transaction is invalidated by other clients to the database.  It is important to note that you cannot preserve transactionality between an `STM` and a `sqlx.Tx`, so any code that must change things in both etcd and postgres must be aware of this (and generally, such code should be avoided).
+
+Operations:
+ * `Get` - same as in `ReadOnlyCollection`, but guaranteed to be consistent with the state of the transaction
+ * `Put` - create or overwrite a given item in the collection
+ * `Update` - get a given item from the collection, modify it in a callback, then write it out afterwards
+ * `Upsert` - same as `Update` except it will continue with the callback and insert if the item does not exist already
+ * `Create` - create a given item in the collection, error if it already exists
+ * `Delete` - remove a given item from the collection, error if it does not exist
+ * `DeleteAll` - remove all items from the collection
+
+`EtcdReadWriteCollection` additional supports:
+ * `TTL` - get the time-to-live of a given item in the collection
+ * `PutTTL` - write an item to the collection that will be removed after the given time-to-live
+
+## Implementation
+
+## Etcd
+
+TODO
+
+### STM
+
+The `STM` can be constructed via `NewSTM` (or `NewDryrunSTM`, which will not commit changes).  This code is copied from the etcd library in order to extend it.  It is a relatively common problem that etcd transactions will get too large and be rejected by etcd.  While the etcd config can be adjusted to increase the limit on transaction size, this can only do so much.  For any arbitrarily-large transactions, you may need to find a way to break it up into pieces of a controlled size.
+
+## Postgres
+
+The postgres implementation is written to mirror the functionality of the etcd implementation rather than to provide an ORM-style table of the protobuf fields inside a database.
+
+### sqlx.Tx
+
+The `sqlx.Tx` can be constructed via `NewSQLTx`.  This will occupy a sql session for its lifetime.  Constructing tehe `sqlx.Tx` will result in a `BEGIN;` at the start, and a `COMMIT;` at the end of the callback, unless an error occurs, in which case a `ROLLBACK;` will be issued.  This will reattempt up to three times in the case of a commit error.
+
+### Tables
+
+A table will be created for each postgres collection.  Construction of the collection will check that the table exists, as well as all relevant triggers and indexes.  If any are missing, they will be created.
```

Comment:
I don't think this is a great idea.  I can imagine difficult to fix bugs coming from lazily creating the tables.  I think we should split that functionality out of the constructor, and have the constructor just fail if the table doesn't exist.  It would be expected that the setup code has already been called before an STM can be created.

Reply:
That sounds fine to me - I suppose I can just move all the initialization into a migration, but it becomes a bit more cumbersome to declare a new collection.